File: src/main/java/com/github/noamm9/packdisabler/mixin/MixinFallbackResourceManager.java
package com.github.noamm9.packdisabler.mixin;

import com.github.noamm9.packdisabler.FilteredPackResources;
import com.github.noamm9.packdisabler.MixinHooks;
import net.minecraft.resources.Identifier;
import net.minecraft.server.packs.PackResources;
import net.minecraft.server.packs.resources.FallbackResourceManager;
import org.spongepowered.asm.mixin.Final;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.ModifyArgs;
import org.spongepowered.asm.mixin.injection.invoke.arg.Args;

import java.util.function.Predicate;

@Mixin(FallbackResourceManager.class)
public class MixinFallbackResourceManager {
    @Shadow @Final private String namespace;

    @ModifyArgs(
        method = {
            "push(Lnet/minecraft/server/packs/PackResources;)V",
            "push(Lnet/minecraft/server/packs/PackResources;Ljava/util/function/Predicate;)V",
            "pushFilterOnly"
        },
        at = @At(
            value = "INVOKE",
            target = "Lnet/minecraft/server/packs/resources/FallbackResourceManager;pushInternal(Ljava/lang/String;Lnet/minecraft/server/packs/PackResources;Ljava/util/function/Predicate;)V"
        )
    )
    private void filterHypixelPack(Args args) {
        if (!MixinHooks.shouldFilterHypixelPack(namespace, args.get(0))) return;

        PackResources pack = args.get(1);
        if (pack != null) args.set(1, new FilteredPackResources(pack));
        else args.set(2, (Predicate<Identifier>) id -> false);
    }
}

File: src/main/kotlin/com/github/noamm9/packdisabler/mixin/MixinFallbackResourceManager.kt
package com.github.noamm9.packdisabler.mixin

import com.github.noamm9.packdisabler.HypixelPackLoader
import com.github.noamm9.packdisabler.config.Config
import net.minecraft.resources.Identifier
import net.minecraft.server.packs.resources.FallbackResourceManager
import net.minecraft.server.packs.resources.Resource
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.Unique
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.*

@Mixin(FallbackResourceManager::class)
abstract class MixinFallbackResourceManager {
    @Shadow abstract fun getResourceStack(location: Identifier): List<Resource>

    @Inject(method = ["listResources"], at = [At("RETURN")], cancellable = true)
    private fun selectTextColorResources(path: String, filter: Predicate<Identifier>, cir: CallbackInfoReturnable<Map<Identifier, Resource>>) {
        val resources = cir.returnValue.toMutableMap()
        resources.keys.filter(::isTextColorResource).forEach { location ->
            val stack = getResourceStack(location)
            val selected = if (Config.newTextColors) {
                stack.find { it.sourcePackId() == HypixelPackLoader.packId } ?: stack.lastOrNull()
            }
            else stack.findLast { it.sourcePackId() != HypixelPackLoader.packId }

            if (selected == null) resources.remove(location)
            else resources[location] = selected
        }

        cir.returnValue = resources
    }

    @Unique
    private fun isTextColorResource(location: Identifier): Boolean {
        if (location.namespace != "minecraft") return false
        return when (location.path) {
            "shaders/core/text.vsh",
            "shaders/core/rendertype_text.vsh",
            "shaders/core/rendertype_text_see_through.vsh",
            "shaders/core/rendertype_text_intensity.vsh",
            "shaders/core/rendertype_text_intensity_see_through.vsh",
            "shaders/include/modify_vanilla_color.glsl",
            "shaders/include/color_util.glsl" -> true

            else -> false
        }
    }
}